File: 5 - Coding Interview Workshops/Week7/CodeFiles/Python/polymorph.py
# Create an interface by not defining all the methods 
class Vehicle:
    def __init__ (self, make, model, year):
        self.make = make
        self.model = model
        self.year = year

    def letsDrive (self):
        pass

class Motorbike (Vehicle):
    def letsDrive (self):
        print("Zooom zooom")

class Car (Vehicle):
    def letsDrive (self):
        print("Vroom vrooooom")

class Truck (Vehicle):
    def letsDrive (self):
        print("Brruuum brrruuuum")

motorbike = Motorbike("Yamaha", "Road Star", 2004)
car = Car("Volkswagen", "Polo", 2006)
truck = Truck("Ford", "F-150", 2021)

vehicles = [motorbike, car, truck]

for vehicle in vehicles:
    vehicle.letsDrive()

    

File: 5 - Coding Interview Workshops/Week07/CodeFiles/JavaScript/polymorph.js
// Interfaces don't exist in JavaScript
// All methods can be overridden and all functions are polymorphic
class Vehicle {
    constructor (make, model, year) {
        this.make = make;
        this.model = model;
        this.year = year;
    }

    letsDrive () {
        console.log("Chug chhuug");
    }
}

class Motorbike extends Vehicle {
    letsDrive () { console.log("Zooom zooom"); }
}

class Car extends Vehicle {
    letsDrive () { console.log("Vroom vrooooom"); }
}

class Truck extends Vehicle {
    letsDrive () { console.log("Brruuum brrruuuum"); }
}

let motorbike = new Motorbike("Yamaha", "Road Star", 2004);
let car = new Car("Volkswagen", "Polo", 2006);
let truck = new Truck("Ford", "F-150", 2021);

let vehicles = [motorbike, car, truck];

vehicles.forEach((vehicle) => vehicle.letsDrive());


